### PR TITLE
feat: Add KubeVirt VirtualMachine custom health checks [#7175]

### LINKED ISSUE #7175: Resource Customizations for KubeVirt Data Types
# Summary

As a KubeVirt user, I would like for Argo-CD to better understand status fields of VirtualMachines, VirtualMachineInstances, and DataVolumes.

# Motivation

DataVolumes often require a long download or PVC clone phase, where they are manifest complete, but not "Conditions=Ready"

VirtualMachines are dependent on creating a VMI, which can only be done once relevant storage is sorted

VirtualMachineInstances are much like Pods, and need to be tracked through resource and node scheduling, instantiation, and running.

# Proposal

I have a set of health.lua scripts ready to go and tested via direct injection into the argocd CR at https://github.com/cwilkers/kubevirt-gitops

I have nearly completed a PR to add these to the resource-customizations sub directory in this repository.

### PR DESCRIPTION
**Pull request by cwilkers**
Signed-off-by: cwilkers <cwilkers@redhat.com>

Checklist:

Fixes #7175  Resource Customizations for KubeVirt Data Types

This does not need to be in release notes nor docs

Red Hat appears in USERS.md

### CODE REVIEW

File: resource_customizations/kubevirt.io/VirtualMachine/health.lua
Comment (zcahana) [inline resource_customizations/kubevirt.io/VirtualMachine/health.lua:1]: The VM state is computed by its controller and is derived from quite a few factors. The VM's `.status.printableStatus` field captures a human-readable summary of this state. We can use the value of this field in the `hs.message`, e.g. `The VirtualMachine is in the '%s' state", obj.status.printableStatus`.

As for the `hs.status`, I think the following is a reasonable mapping to Argo CD [health states](https://github.com/argoproj/gitops-engine/blob/master/pkg/health/health.go):

- !created --> Suspended
- created && Paused --> Suspended
- created && Ready --> Healthy
- created && !Ready --> Progressing

As for the last one, it isn't always accurate since some states better map to Degrading (e.g., various error states), but printableStatus in the message can be used to convey that extra information.

File: resource_customizations/kubevirt.io/VirtualMachine/health.lua
Comment (zcahana) [inline resource_customizations/kubevirt.io/VirtualMachine/health.lua:4]: I guess you wanted to set `hs.message` here?

File: resource_customizations/kubevirt.io/VirtualMachineInstance/health.lua
Comment (zcahana) [inline resource_customizations/kubevirt.io/VirtualMachineInstance/health.lua:1]: For a VMI, there's a stable [`.status.phase`](https://github.com/kubevirt/client-go/blob/148fa0d1c7e83b7a56606a7ca92394ba6768c9ac/api/v1/types.go#L549) field which can be used (alongside the Ready and Paused conditions).
Here's a reasonable mapping to Argo CD health states:
- Pending, Scheduling, Scheduled --> Progressing
- Succeeded --> Suspended (can also use Healthy but I think Suspended conveys the more significant aspect here)
- Failed --> Degraded (can also use Suspended and have the `hs.message` discriminate between Succeeded/Failed/Paused)
- Unknown --> Unknown
- Running -->
    - Paused --> Suspended
    - !Ready --> Degraded
    - Ready --> Healthy

File: resource_customizations/cdi.kubevirt.io/DataVolumes/health.lua
Comment (zcahana) [inline resource_customizations/cdi.kubevirt.io/DataVolumes/health.lua:1]: We can supplement the phase with the Running condition to try and expose errors:
- Succeeded --> Healthy
- Failed --> Degraded
- Unknown --> Unknown
- Paused --> Suspended
- Other phase + Running --> Progressing
- Other phase + !Running --> Degraded

File: resource_customizations/kubevirt.io/VirtualMachine/health.lua
Comment (cwilkers) [reply resource_customizations/kubevirt.io/VirtualMachine/health.lua]: Oops, thanks, I just force-pushed a quick update for this.

File: resource_customizations/kubevirt.io/VirtualMachine/health.lua
Comment (cwilkers) [reply resource_customizations/kubevirt.io/VirtualMachine/health.lua:1]: For every permutation of status and message,  I will need to write a test with an example manifest.
My original thought was to follow the `.status.conditions` but this only updates IFF the VM is running.

I like the idea of providing some outputs to Suspended, as that could provide useful information where manual intervention is needed to get an application to progress. I will experiment with this more.

Comment (rmohr) [reply resource_customizations/kubevirt.io/VirtualMachine/health.lua:1]: The  suggestion looks great, one point in addition:

> * created && !Ready --> Progressing

If a VM would be come unhealthy it would go back to `Progressing` instead of `Degraded`. I am not sure if one can detect right now immediately on the status if the VM was ready before and became unready.

One way could work like this:

 * We safe the phase transition timestamps on the VMIs. One could look up the initial start delay for readiness probes, take the running transition timestamp and show "progressing" until timestamp + readiness delay is over and from that timepoint on always show "degraded".

Comment (zcahana) [reply resource_customizations/kubevirt.io/VirtualMachine/health.lua:1]: > I am not sure if one can detect right now immediately on the status if the VM was ready before and became unready.

You're right, in this case the VM's printableStatus would still show Running, and alongside a Progressing argocd status - we're missing an adequate indication that something's wrong with the VM.

> We safe the phase transition timestamps on the VMIs

The context in which this lua script operates, have access only to the VM object, and can't (and, IMO, shouldn't) query the kubernetes API to fetch the corresponding VMI.

Another option is to look at the VM's printableStatus, and if it's Running and !Ready, then set a Degraded status. Can also do that for a Migrating (&& !Ready) printableStatus.
True, this violates our intention of not to write automation based on the printableStatus since its states might be changed/added/removed. But if we consider that the Running state is not likely to ever disappear, this can probably present more value than risk.

Comment (cwilkers) [reply resource_customizations/kubevirt.io/VirtualMachine/health.lua:1]: I just pushed updates to the VM health check. I center on the status.printableStatus field, with a simplistic fallback in case of an old version missing the field.

Comment (rmohr) [reply resource_customizations/kubevirt.io/VirtualMachine/health.lua:1]: Hm, `status.printableStatus` is a human-readable only field. It can change anytime and should not be relied on.

Comment (rmohr) [reply resource_customizations/kubevirt.io/VirtualMachine/health.lua:1]: If I am not mistaken we already mirror the readiness probe to the VM too. That should be enough to find out about degraded states.

Comment (cwilkers) [reply resource_customizations/kubevirt.io/VirtualMachine/health.lua:1]: @rmohr I've made an adjustment here to set Healthy based on `status.ready` instead of printableStatus.

The other states default to Progressing and printableStatus, so I think we are okay even if the style of printableStatus changes. The only status we want to get 100% correct is Healthy.

File: resource_customizations/kubevirt.io/VirtualMachine/health.lua
Comment (zcahana) [inline resource_customizations/kubevirt.io/VirtualMachine/health.lua:3]: Note that `.status.ready` and `.status.created` are boolean field, so you should actually be looking at their values, and not their presence.

File: resource_customizations/kubevirt.io/VirtualMachine/health.lua
Comment (zcahana) [inline resource_customizations/kubevirt.io/VirtualMachine/health.lua:11]: To minimize reliance on the printableStatus field, we can use the following:
- For "Stopped", you can check that `created == false`.
- For "Paused", you can check the condition with type "Paused" and value "True".

File: resource_customizations/kubevirt.io/VirtualMachineInstance/health.lua
Comment (zcahana) [inline resource_customizations/kubevirt.io/VirtualMachineInstance/health.lua:14]: This should also set to Degraded if !Ready, and would also be nice to set to Suspended when the Paused condition is set to True.

File: resource_customizations/cdi.kubevirt.io/DataVolume/health.lua
Comment (zcahana) [inline resource_customizations/cdi.kubevirt.io/DataVolume/health.lua:11]: In addition to the phase, it may be worthwhile to look at the `Running` condition, and report `Progressing` only if the condition is true. For example, a DataVolume which points to a non-existing source, will be stuck forever in an `ImportInProgress` phase, and its importer pod will continuously crash-loop. Most of the time, the `Running` condition will be set to false (except, momentarily when the importer pod is relaunched, but due to the exponential backoff it'll happen less and less often).

**Comment by cwilkers** [comment]
> Side question that shouldn't necessarily impact the merge: @cwilkers what would be the best way to test this out end-to-end? Would you have sample CRs I could use to try this out on OpenShift ?

Yes! The basic examples I have been using are from [KubeVirt's labs](https://kubevirt.io/labs/kubernetes/lab2.html).

The DataVolume can be created as soon as KubeVirt is up and running (lab1 on that site) Then there is a VirtualMachine to add, and the VirtualMachineInstance is created from the VM once it gets scheduled on a node. In normal usage, you would never directly create a VMI resource, only a VM or DV.

**Review by jessesuen — COMMENTED** [review]
I took a look at this PR since it got put onto the Argo CD maintainer's agenda. I'm not familiar with Kubevirt, but it looks like @zcahana is, based on GitHub activity, so I'm trusting his feedback on this PR and would like him to sign off on it before merging.

**Review by jessesuen — APPROVED** [review]

**Pull request merged**
